portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -932,21 +1003,125 @@ protected String generateToString() {
 			}
 		}
 
-		Map<String, String[]> renderParams = _params;
+		Map<String, String[]> portletURLParams = new LinkedHashMap<>();
 
-		if (_copyCurrentRenderParameters &&
-			!(_lifecycle.equals(PortletRequest.RESOURCE_PHASE) &&
-			  _cacheability.equals(ResourceURL.FULL))) {
+		Set<String> actionParameterNames;
 
-			renderParams = _mergeWithRenderParameters(renderParams);
+		if (_mutableActionParameters != null) {
+			actionParameterNames = _mutableActionParameters.getNames();
+
+			for (String parameterName : actionParameterNames) {
+				portletURLParams.put(
+					_ACTION_PARAMETER_NAMESPACE + parameterName,
+					_mutableActionParameters.getValues(parameterName));
+			}
+		}
+		else {
+			actionParameterNames = Collections.emptySet();
 		}
 
-		for (Map.Entry<String, String[]> entry : renderParams.entrySet()) {
-			String name = entry.getKey();
+		Set<String> resourceParameterNames = Collections.emptySet();
 
-			if (isParameterIncludedInPath(name)) {
-				continue;
+		if (_mutableResourceParameters != null) {
+			resourceParameterNames = _mutableResourceParameters.getNames();
+
+			for (String parameterName : resourceParameterNames) {
+				portletURLParams.put(
+					_RESOURCE_PARAMETER_NAMESPACE + parameterName,
+					_mutableResourceParameters.getValues(parameterName));
 			}
+		}
+
+		if (!_lifecycle.equals(PortletRequest.RESOURCE_PHASE) ||
```

Comment:
@shuyangzhou Directing your attention to line 1035 -- I would say that this was the most difficult part of the `generateToString()` method to get correct, because it deals with appending both public and private render parameters to the `String` return value. *However*, it rather than appending them immediately, instead it builds up the key+value pairs in a temporary `portletURLParams` map.

*Terminology/Background:*

The Portlet 3.0 API and Spec prefer to use the term "private render parameters" instead of the old Portlet 2.0 term "render parameters" -- they mean the same thing, it's just that Portlet 3.0 was trying to make the language more clear by introducing the word "private". I considered renaming our `RenderParametersPool` class to `PrivateRenderParametersPool` but I didn't want to change the API too much.

The Spec and API Javadoc have many rules and requirements for appending both public and private render parameters to the `String` return value. The code here starting at line 1035 implements those rules, but as I wrote earlier, rather than appending them immediately, instead it builds up the key+value pairs in a temporary `portletURLParams` map. Appending happens further down.